src/std_extensions/arithmetic/conversions/const_fold.rs

```diff
@@ -0,0 +1,134 @@
+use crate::{
+    extension::{
+        prelude::{sum_with_error, ConstError},
+        ConstFold, ConstFoldResult, OpDef,
+    },
+    ops,
+    std_extensions::arithmetic::{
+        float_types::ConstF64,
+        int_types::{get_log_width, ConstIntS, ConstIntU, INT_TYPES},
+    },
+    types::ConstTypeError,
+    values::{CustomConst, Value},
+    IncomingPort,
+};
+
+use super::ConvertOpDef;
+
+pub(super) fn set_fold(op: &ConvertOpDef, def: &mut OpDef) {
+    use ConvertOpDef::*;
+
+    match op {
+        trunc_u => def.set_constant_folder(TruncU),
+        trunc_s => def.set_constant_folder(TruncS),
+        convert_u => def.set_constant_folder(ConvertU),
+        convert_s => def.set_constant_folder(ConvertS),
+    }
+}
+
+fn get_input<T: CustomConst>(consts: &[(IncomingPort, ops::Const)]) -> Option<&T> {
+    let [(_, c)] = consts else {
+        return None;
+    };
+    c.get_custom_value()
+}
+
+fn fold_trunc(
+    type_args: &[crate::types::TypeArg],
+    consts: &[(IncomingPort, ops::Const)],
+    convert: impl Fn(f64, u8) -> Result<Value, ConstTypeError>,
+) -> ConstFoldResult {
+    let f: &ConstF64 = get_input(consts)?;
+    let f = f.value();
+    let [arg] = type_args else {
+        return None;
+    };
+    let log_width = get_log_width(arg).ok()?;
+    let int_type = INT_TYPES[log_width as usize].to_owned();
+    let sum_type = sum_with_error(int_type.clone());
+    let err_value = || {
+        let err_val = ConstError {
+            signal: 0,
+            message: "Can't truncate non-finite float".to_string(),
+        };
+        let sum_val = Value::Sum {
+            tag: 1,
+            value: Box::new(err_val.into()),
+        };
+
+        ops::Const::new(sum_val, sum_type.clone()).unwrap()
+    };
+    let out_const: ops::Const = if !f.is_finite() {
+        err_value()
+    } else {
+        let cv = convert(f, log_width);
+        if let Ok(cv) = cv {
+            let sum_val = Value::Sum {
+                tag: 0,
+                value: Box::new(cv),
+            };
+
+            ops::Const::new(sum_val, sum_type).unwrap()
+        } else {
+            err_value()
+        }
+    };
+
+    Some(vec![(0.into(), out_const)])
+}
+
+struct TruncU;
+
+impl ConstFold for TruncU {
+    fn fold(
+        &self,
+        type_args: &[crate::types::TypeArg],
+        consts: &[(IncomingPort, ops::Const)],
+    ) -> ConstFoldResult {
+        fold_trunc(type_args, consts, |f, log_width| {
+            ConstIntU::new(log_width, f.trunc() as u64).map(Into::into)
```

Comment:
What if `f` is negative? Does `as u64` cause a panic?

Reply:
`as u64` maps all negative floats to 0, should I add a panic?

Reply:
Hmm, this is a can of worms. The spec for `trunc_u` currently says "Returns an error when the float is non-finite or cannot be exactly stored in N bits". This should probably say "cannot be exactly represented as a `u<N>`". But what we're doing here is rounding, which is much more forgiving. WASM [seems to say](https://webassembly.github.io/spec/core/bikeshed/#-hrefop-trunc-umathrmtruncmathsfu_m-n-z) that the result is undefined if the number is negative, but rounded otherwise, which makes me think we should change the definition of `trunc_{u,s}` in the spec.

Reply:
sounds like this could be a follow up issue?

Reply:
Yes, I'll make one.